File: 482. License Key Formatting/solution.py
class Solution(object):
    def licenseKeyFormatting(self, S, K):
        """
        :type S: str
        :type K: int
        :rtype: str
        """
        sol = []
        count = 0
        for i in range(len(S)-1, -1, -1):
            s = S[i]
            if s == '-':
                continue
            elif 48 <= ord(s) <= 57:
                sol.append(s)
                count += 1
            elif 97 <= ord(s) <= 122:
                sol.append(chr(ord(s)-32))
                count += 1
            else:
                sol.append(s)
                count += 1
            if count != 0 and count%K == 0:
                sol.append("-")

        if sol and sol[-1] == '-':
            sol = sol[:-1]
        
        return "".join(sol[::-1])
            
        

File: 482. License Key Formatting/solution.java
class Solution {
    public String licenseKeyFormatting(String S, int K) {
        StringBuilder sol  = new StringBuilder();
        StringBuilder batch;
        int n = S.length();
        
        Character c;
        int i = n - 1;
        while (i >= 0) {
            batch = new StringBuilder();
            int j = 0;
            while (j < K && i >= 0) {
                if (S.charAt(i) != '-') {
                    c = S.charAt(i);
                    if (Character.isLowerCase(c)) {
                        c = Character.toUpperCase(c);
                    }
                    j++;
                    batch.append(c);
                }
                i--;

            }
            if (batch.length() != 0 && sol.length() != 0) {
                sol.append(new String("-"));
            }
            sol.append(batch.toString());
        }
        sol.reverse();
        return sol.toString();
    }
}
